src/hotspot/share/gc/shared/stringdedup/stringDedupTable.cpp

```diff
@@ -395,32 +395,33 @@ bool StringDedupTable::is_rehashing() {
 StringDedupTable* StringDedupTable::prepare_resize() {
   size_t size = _table->_size;
 
-  // Check if the hashtable needs to be resized
+  // Decide whether to resize, and compute desired new size if so.
   if (_table->_entries > _table->_grow_threshold) {
-    // Grow table, double the size
-    size *= 2;
-    if (size > _max_size) {
-      // Too big, don't resize
-      return NULL;
+    // Compute new size.
+    size_t needed = _table->_entries / _grow_load_factor;
+    if (needed < _max_size) {
+      size = round_up_power_of_2(needed);
+    } else {
+      size = _max_size;
     }
   } else if (_table->_entries < _table->_shrink_threshold) {
-    // Shrink table, half the size
-    size /= 2;
-    if (size < _min_size) {
-      // Too small, don't resize
-      return NULL;
-    }
-  } else if (StringDeduplicationResizeALot) {
-    // Force grow
-    size *= 2;
-    if (size > _max_size) {
-      // Too big, force shrink instead
-      size /= 4;
+    // Compute new size.  We can't shrink by more than a factor of 2,
+    // because the partitioning for parallelization doesn't support more.
+    if (size > _min_size) size /= 2;
```

Comment:
I would prefer braces :) But an even nicer solution would be to use `clamp(...)` both for the grow and shrink case. What do you think about that?

Reply:
I think clamp just confuses things because only one of the bounds matters,
which bound depending on whether growing or shrinking. And on the grow side
the prior test against _max_size is still needed to eliminate any questions
about overflow. That is, I wouldn't want to just use
```
size = clamp(round_up_power_of_2(needed), _min_size, _max_size);
```
without the prior check, because the round_up could perhaps overflow (at
least in theory, and maybe only on 32bit platforms).  But with the prior
check the clamp is superfluous.

Reply:
I see, being sure there is no overflow is good (even if it feels very theoretical in this case). Otherwise I was thinking something like: 
```
if (grow) {
  size = round_up_power_of_2(needed);
} else if (shrink) {
  size = size / 2;
}
clamp(size, _min_size, _max_size);
```